### ISSUE TITLE
issue on x64

### ISSUE THREAD

**Issue by winpassuser** [Snowapril/String-Obfuscator-In-Compile-Time#1]
In the following code I had to change "unsigned int..." into "size_t..." to make it compile with gcc 7.4 on x64. The obfuscation does not happen with the default settings, but surprisingly still works with -O3.

`template <unsigned int... I, int A, int B>
class MetaString<std::index_sequence<I...>, A, B> {`

**Comment by Snowapril** [comment]
* In the following code I had to change "unsigned int..." into "size_t..." to make it compile with gcc 7.4 on x64. 

I checked the problem what you commented and fixed it. Thank you.


* The obfuscation does not happen with the default settings, but surprisingly still works with -O3.

At first, I tried to fix the problem about "**constexpr evaluation does not occurred in the default optimization level (by default -O0 in g++)**", but I could not fix it. Therefore, I found a similar issue in the stackoverflow. I quote a proper answer to our issue. (reference : https://stackoverflow.com/questions/41862118/why-gcc-does-not-evaluate-constexpr-at-compile-time)

> In other situations, even when a constexpr function is invoked with arguments that are all themselves constant expressions, it is up to the implementation to do what it wants. Typically, it'll depend on the optimization flags. **On both gcc 6.2 and clang 3.9.1, for instance, -O0 will emit a call at runtime but -O1 will emit the constant 36000**  </br>

In summary, gcc with default optimization level evaluates the constexpr expressions in runtime, but -O1 or higher levels evaluate them in compile time. Therefore, an original string can be exposed to the disassembled code in the default optimization level. In practice, I think this is not a problem because there is no one distributing their product with the default optimization level.

In our case, It worked properly in -O1 or higher optimization levels.

**Comment by Snowapril** [comment]
This issue will be closed in a week. Thank you.
